- ####在JAVA中，用類的形式對不正常情況進行了描述和封裝對象。
- ####描述不正常情況的類，就稱為異常類。
- ####異常就是JAVA通過面向對象的思想將問題封裝成了對象，提高閱讀性。
- ####用異常類對其進行描述，不同的問題用不同的類進行具體描述。

##異常(throwable):
    1. Error (錯誤)      錯誤一班都由於jvm或者是硬體引發的問題,我們一般都不會通過代碼處理
    2. Exception (異常)  是在運行時期發生的不正常情況，異常我們一般都會通過代碼處理

##異常的分類：
    1. 編譯時被檢測異常：只要是Exception和其子類都是，除了特殊子類RuntumeException體系。
        這種問題一旦出現，希望在編譯時就進行檢測，讓這種問題有對應的處理方式。
        這樣的問題都可以針對性的處理。
        
    2.  編譯時不檢測異常（運行時異常）：就是Exception中的RuntimeException和其子類。
        這種問題的發生，無法讓功能繼續，運算無法進行，更多是因為調用者的原因導致的，或者引發了內部狀態的改變導致的。
        那麼這問題一般不處理，直接編譯通過，在運行時，讓調用者的程序強制停止，讓調用者對代碼進行修正。
        此類異常,在方法聲明時可不必聲明拋出(throws)
    
    所以自定義異常時，要麼繼承Exception,要麼繼承RuntimeException。

- throw用於函數主體,throws用於函數宣告
- throws拋出的是異常類，可拋出多個，用逗號隔開。
- throw拋出的是異常對象。
異常處理方式

   > 一. 捕獲處理
        try{
            或許會發生問題的代碼..
        } catch(捕獲的異常類型 變量名){
            處理異常的代碼...
        }
    >>捕獲處理注意事項:
            1.如果一個try塊中出現異常代碼,經過處理之後,try-catch塊外面的代碼可繼續執行
            2.如果一個try塊中出現了異常代碼,在try塊中出現異常的代碼後面的語句無法執行
            3.一個try塊後面可以跟多個catch塊,也就是一個try塊可以捕獲多種異常的類型,但是捕獲的異常類型必須從小到大進行捕獲

   > 二. 拋出處理(throw throws)
        拋出處理注意細節:
            1.如果一個方法內部拋出一個異常現象,那麼該方法要聲明拋出
            2.如果調用一個聲明拋出的方法,調用者必須要處理
            3.如果一個方法拋出一個異常對象,該方法就會馬上停止(一個方法遇到throw關鍵字,那麼該方法就會停止)
            4.在一種狀況下只會拋出一種異常
   
   try catch finally 代碼塊組合特點:
    > 一.    try catch finally
    > 二.    try catch(多catch) 當沒有必要資源釋放時，可以不用定義finally。
    > 三.    try finally 異常無法直接catch處理，但是資源需要關閉。
            void show()throws Exception  //必須聲明拋出,因為無catch處理
            {
                try
                {
                    //開啟資源。
                    new Exception();
                }
                finally
                {
                    //關閉資源。
                }
            }

    異常的注意事項：
    1.  子類在覆蓋父類方法時，父類的方法如果拋出了異常，
        那麼子類的方法只能拋出父類的異常，或者該異常的子類。
    
    2.  如果父類拋出多個異常，那麼子類只能拋出父類異常的子集。
    
    簡單說：子類覆蓋父類只能拋出父類的異常或者子類或者子集。
    注意： 如果父類的方法沒有拋出異常，那麼子類覆蓋時絕對不能拋。(只能try.)
